src/internal_events/pulsar.rs

```diff
@@ -56,3 +58,105 @@ impl<F: std::fmt::Display> InternalEvent for PulsarPropertyExtractionError<F> {
         );
     }
 }
+
+#[derive(Debug)]
+pub struct PulsarReadError {
+    pub error: pulsar::Error,
+}
+
+impl InternalEvent for PulsarReadError {
+    fn emit(self) {
+        error!(
+            message = "Failed to read message.",
+            error = %self.error,
+            error_code = "reading_message",
+            error_type = error_type::READER_FAILED,
+            stage = error_stage::RECEIVING,
+            internal_log_rate_limit = true,
+        );
+        counter!(
+            "component_errors_total", 1,
+            "error_code" => "reading_message",
+            "error_type" => error_type::READER_FAILED,
+            "stage" => error_stage::RECEIVING,
+        );
+        // deprecated
+        counter!("events_failed_total", 1);
+    }
+}
+
+#[derive(Debug)]
+pub struct PulsarAcknowledgmentError {
+    pub error: ConsumerError,
+}
+
+impl InternalEvent for PulsarAcknowledgmentError {
+    fn emit(self) {
+        error!(
```

Comment:
For all metrics, I propose you use the "new" metrics interface. Here is an example of this interface: https://github.com/vectordotdev/vector/blob/master/src/sources/amqp.rs#L336 You can check other places for more examples. This new pattern should be used for all metrics since it has much less performance overhead.

If you still have questions about this metrics approach, I believe @bruceg can answer them.

Reply:
Oh, I will check it tomorrow.

Reply:
I introduced `PulsarErrorEvent` to replace the "old" metrics.  https://github.com/vectordotdev/vector/pull/18475/commits/0af4d1e4e19715872d300c32384c7154e22db50f
For ease of implementation, I used the original type `EventsReceived` directly instead of converting `PulsarEventsReceived` to topic-level "new" type which need to maintain dynamic pulsar topics. https://github.com/vectordotdev/vector/pull/18475/commits/38f17de94de6e5bd781cc845e951123594b0b442